apm-agent-plugins/apm-java-concurrent-plugin/src/main/java/co/elastic/apm/agent/concurrent/JavaConcurrent.java

```diff
@@ -0,0 +1,229 @@
+     *
+     * <p>
+     * Note: does activates the {@link AbstractSpan} and not only the {@link TraceContext}.
+     * This should only be used when the span is closed in thread the provided {@link Runnable} is executed in.
+     * </p>
+     */
+    @Nullable
+    public static <T> Callable<T> withContext(@Nullable Callable<T> callable, @Nullable ElasticApmTracer tracer) {
+        if (callable instanceof CallableLambdaWrapper || callable == null || tracer == null  || needsContext.get() == Boolean.FALSE) {
+            return callable;
+        }
+        needsContext.set(Boolean.FALSE);
+        AbstractSpan<?> active = tracer.getActive();
+        if (active == null) {
+            return callable;
+        }
+        if (isLambda(callable)) {
+            callable = new CallableLambdaWrapper<>(callable);
+        }
+        ElasticApmAgent.ensureInstrumented(callable.getClass(), RUNNABLE_CALLABLE_FJTASK_INSTRUMENTATION);
+        contextMap.put(callable, active);
+        active.incrementReferences();
+        return callable;
+    }
+
+    @Nullable
+    public static <T> ForkJoinTask<T> withContext(@Nullable ForkJoinTask<T> task, @Nullable ElasticApmTracer tracer) {
+        if (task == null || tracer == null  || needsContext.get() == Boolean.FALSE) {
+            return task;
+        }
+        needsContext.set(Boolean.FALSE);
+        AbstractSpan<?> active = tracer.getActive();
+        if (active == null) {
+            return task;
+        }
+        ElasticApmAgent.ensureInstrumented(task.getClass(), RUNNABLE_CALLABLE_FJTASK_INSTRUMENTATION);
+        contextMap.put(task, active);
+        active.incrementReferences();
+        return task;
+    }
+
+    public static void doFinally(@Nullable Throwable thrown, @Nullable Object contextObject) {
+        needsContext.set(Boolean.TRUE);
+        if (thrown != null && contextObject != null) {
+            removeContext(contextObject);
+        }
+    }
+
+    public static void doFinally(@Nullable Throwable thrown, @Nullable Collection<? extends Callable<?>> callables) {
+        needsContext.set(Boolean.TRUE);
+        if (thrown != null && callables != null) {
+            for (Callable<?> callable : callables) {
+                removeContext(callable);
+            }
+        }
+    }
+
+    private static boolean isLambda(Object o) {
+        return o.getClass().getName().indexOf('/') != -1;
+    }
+
+    @Nullable
+    public static <T> Collection<? extends Callable<T>> withContext(@Nullable Collection<? extends Callable<T>> callables, @Nullable ElasticApmTracer tracer) {
+        if (callables == null || tracer == null) {
+            return null;
+        }
+        if (callables.isEmpty()) {
+            return callables;
+        }
+        final Collection<Callable<T>> wrapped;
+        if (needsWrapping(callables)) {
+            wrapped = new ArrayList<>(callables.size());
```

Comment:
This introduces a risk of causing `ClassCastException` due to the collection type. I think it's better to try/catch using `Collection.remove` and `Collection.add` and avoid wrapping (with logging) if not supported by the collection.

Reply:
That would impact the order of the callables in the collection. But we could clear and re-fill it.

Reply:
I think calling `add` and `remove` is more likely to fail than creating an entirely new list because that's a good chance the collection is immutable.
Maybe leave as-is and remove if we see it causes problems?

Reply:
My vote is actually to go the other way around - fail (meaning - don't wrap lambdas in collections if the collection is immutable) or not try to wrap in collections and see if we get requests to support that. I prefer questions about why something is not working than issues with breaking applications.
If you are not convinced - leave as is.

Reply:
That may cause subtle context losses that are very hard to debug. I'd like to resort to these defensive measures only if there are actual problems in the wild which will surface in a more obvious way by throwing an exception. I'll add a new instrumentation group name `executor-collection` to disable just this instrumentation more easily.

I think in practice it's unlikely that an executor implementation depends on the type of the collection or even overrides the default implementation in `AbstractExecutorService`.

Reply:
This entire PR is about executors exposing APIs that allow interfaces but expect specific types 🙂